src/ModelContextProtocol/Protocol/Messages/ProgressNotification.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System.ComponentModel;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace ModelContextProtocol.Protocol.Messages;
+
+/// <summary>
+/// An out-of-band notification used to inform the receiver of a progress update for a long-running request.
+/// <see href="https://github.com/modelcontextprotocol/specification/blob/main/schema/">See the schema for details</see>
+/// </summary>
+[JsonConverter(typeof(Converter))]
+public class ProgressNotification
+{
+    /// <summary>
+    /// The progress token which was given in the initial request, used to associate this notification with the request that is proceeding.
+    /// </summary>
+    public required ProgressToken ProgressToken { get; init; }
+
+    /// <summary>
+    /// The progress thus far. This should increase every time progress is made, even if the total is unknown.
+    /// </summary>
+    public required ProgressNotificationValue Progress { get; init; }
+
+    /// <summary>Provides a <see cref="JsonConverter"/> for <see cref="ProgressNotification"/>.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class Converter : JsonConverter<ProgressNotification>
+    {
+        /// <inheritdoc />
+        public override ProgressNotification? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            ProgressToken? progressToken = null;
+            float? progress = null;
+            float? total = null;
+            string? message = null;
+            
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
```

Comment:
What necessitates writing a custom converter for this type? If it's because we want to nest information inside `ProgressNotificationValue` couldn't this just be exposed as a JsonIgnore'd getter that synthesizes that instance for you on demand?

Reply:
ProgressNotification is a token and a value, but the value is splatted out flat. When raising progress notifications, I wanted it to be as obvious as possible what the connection is, that you're pro using the values associated with those notifications. I started with it duplicated, but didn't like how unrelated they were. I can go back to it if you feel strongly about it.

Reply:
The reasons I already enumerated. `IProgress<T>` needs to take a single `T`. It can't be a ` ProgressNotification`, because that requires a `ProgressToken` to be supplied, but everything else in the `ProgressNotification` is the `T`. Hence pulling all but the `ProgressToken` out into a `ProgressNotificationValue`, so that a `ProgressNotification` is just effectively a key/value pair of `ProgressToken`/`ProgressNotificationValue`, and then `IProgress<ProgressNotificationValue>` works, anything further added as part of the value in the future "just works" (the message was just added to the spec last week), there's an obvious connection between when a `ProgressNotificationValue` is relative to a `ProgressNotification`, both in name and in structure, etc.